src/poetry/core/packages/utils/utils.py

```diff
@@ -172,9 +178,10 @@ def add_constraint(
     for group_name in requirements:
         # remove duplicates
         seen = []
-        requirements[group_name] = [
-            r for r in requirements[group_name] if not (r in seen or seen.append(r))
-        ]
+        for r in requirements[group_name]:
```

Comment:
Is there a typing reason we can't use the list comprehension here?

Reply:
yeah, it's a bit annoying but it complains that `seen.append(r)` does not return a value

Reply:
I'm not thrilled about that as I find it less readable, but hopefully mypy matures and a future refactor cleans this up. I for one am always rewriting loops as comprehensions where it adds readability, because it's (very moderately) more performant as well.